Approach.md

```diff
@@ -0,0 +1,152 @@
+# Scope
+
+As per the challenge requirements I'll here briefly describe high level design choices, scope, tradeoffs and edgecases.
+
+I'll be aiming to implement the 5th level of the challenge.
+
+From a high level perspective I'll be implementing the following components:
+
+- A service/daemon that implements an async gRPC server with authentication and authorization. From here on referred to as `rrockerd`
+- A worker library that can start/stop/query/stream fully isolated tasks with resource control. This will be a module of `rrockerd` as it's specialized enough that it's unlikely to be useful as a standalone library.
+- A simple CLI with commands to start/stop/query/stream tasks using `rrockerd`. From here on reffered to as `rrocker-cli`
+- A shared gRPC api that will be a standalone library that's easy to consume from both `rrockerd` and `rrocker-cli`. From here on referred to as `rrocker-lib`
+
+Here's a breakdown of the engineering tradeoffs for each component:
+
+## rrockerd:
+I'll be using the rust crate [nix](https://github.com/nix-rust/nix) extensively for the low level systemcalls. Since nix is merely a safe wrapper around libc I believe this is in the spirit of the challenge.
+
+The daemon will be run entirely in memory and will not persist any state in order to survive reboots/crashes.
+
+The gRPC API is assumed to be internal only, as such there won't be any rate limiting, user action logging or abuse mitigations.
+
+Dealing with tty, ptty, terminals and what not is entirely out of scope and as such any program output will be streamed as the raw bytes written to the stdout and stderr pipes.
+Furthermore the output will be chunked on newlines and any lines longer than max gRPC packet size will be truncated.
+
+Resource limits will be implemented using separate cgroups for each
+
+Isolation will be hardcoded to fully isolate each scheduled task from other tasks and the host system. That means a separate PID, Mount, User and Network namespace for each task with no way to connect them.
+There will be no network devices attached to any tasks.
+
+Mounting is also considered out of scope and as such each task will be chrooted to a RW root that's a copy of a base image that's removed upon exit of the task.
+
+I'll make no attempt at Linux backwards compatibility and the minimum kernel version will be 5.0+.
+
+Tasks will inherit the UID/GID of the daemon and internally be mapped as root.
+
+For the gRPC implementation I'll be using the rust crate [tonic](https://github.com/hyperium/tonic) to avoid reinventing the wheel.
+
+Mutable interaction with tasks will be mutually exclusive meaning that the stop action will take a write-only lock on the task such that any concurrent stop task will wait for the first to complete and the subsequently fail as the task has already stopped.
+Read-only interactions such as query and stream can happen concurrently without issue.

```

Comment:
Can describe how you plan to handle streaming?

Reply:
Done! Let me know if you have any further questions.